File: tools/spectra.py
class Spectra:
    
    def __init__(self, ncomps=0, ntrans=0, activity=None, error=None):
        self.ncomps = ncomps
        self.ntrans = ntrans

        if activity is None:
            self.activity = ntrans * ncomps * [0];
        else:
            self.activity = activity

        if error is None:
            self.error = ntrans * [0]
        else:
            self.error = error

    def set(self, comp, trans, val=1):
        self.activity[(trans - 1) * self.ncomps + (comp - 1)] = val

    def set_error(self, trans, val=1):
        self.error[(trans - 1)] = val
    
    def get(self, comp, trans):
        return self.activity[(trans - 1) * self.ncomps + (comp - 1)]

    def get_error(self, trans):
        return self.error[(trans - 1)]

    def write(self):
        print self.ncomps, self.ntrans
        for transaction in range(1, self.ntrans + 1):
            for component in range(1, self.ncomps + 1):
                print self.get(component, transaction),
            print int(self.get_error(transaction))

    def read(self, file):
        buffer = str()
        for line in file:
            buffer += " "+line

        buffer = buffer.split();
        
        self.ncomps = int(buffer[0])
        self.ntrans = int(buffer[1])
        
        self.activity = self.ntrans * self.ncomps * [0]
        self.error = self.ntrans * [0]
        
        i = 2
        for transaction in range(1, self.ntrans + 1):
            for component in range(1, self.ncomps + 1):
                self.set(component, transaction, int(buffer[i]))
                i = i + 1
            self.set_error(transaction, buffer[i] == '1' or buffer[i] == 'x' or buffer[i] == '-');
            i = i + 1

        return self

    def is_candidate(self, candidate):
        for transaction in range(1, self.ntrans + 1):
            if not self.get_error(transaction) > 0:
                continue
            hit = False
            for component in candidate:
                if(self.get(component, transaction) > 0):
                    hit = True
                    break
            if not hit:
                return False
        return True

File: src/common/diagnosis/spectra.cpp
#include "spectra.h"

namespace diagnosis {
bool t_spectra::is_candidate (const t_candidate & candidate,
                              const t_spectra_filter * filter) const {
    t_spectra_iterator it(get_component_count(),
                          get_transaction_count(),
                          filter);


    while (it.next_transaction()) {
        bool hit = false;

        if (!is_error(it.get_transaction()))
            continue;

        t_candidate::iterator candidate_it = candidate.begin();

        while (candidate_it != candidate.end())
            if (get_count(*(candidate_it++), it.get_transaction())) {
                hit = true;
                break;
            }

        if (!hit)
            return false;
    }

    return true;
}

std::ostream & t_spectra::write (std::ostream & out,
                                 const t_spectra_filter * filter) const {
    throw e_not_implemented();
}

std::istream & t_spectra::read (std::istream & in) {
    throw e_not_implemented();
}

std::istream & operator >> (std::istream & in, t_spectra & spectra) {
    return spectra.read(in);
}

std::ostream & operator << (std::ostream & out, const t_spectra & spectra) {
    return spectra.write(out);
}

t_basic_spectra::t_basic_spectra () {
    set_transaction_count(0);
    set_component_count(0);
}

t_basic_spectra::t_basic_spectra (t_count component_count,
                                  t_count transaction_count) {
    set_transaction_count(transaction_count);
    this->transaction_count = transaction_count;
    set_component_count(component_count);
}

t_count t_basic_spectra::get_error_count (const t_spectra_filter * filter) const {
    // FIXME: Improve performance
    t_count total_errors = 0;


    if (filter) {
        for (t_id i = 1; i <= get_transaction_count(); i++)
            if (is_error(i) && !filter->is_transaction(i))
                total_errors++;
    }
    else {
        for (t_id i = 1; i <= get_transaction_count(); i++)
            if (is_error(i))
                total_errors++;
    }

    return total_errors;
}

t_count t_basic_spectra::get_component_count (const t_spectra_filter * filter) const {
    assert(!filter || filter->get_last_component() <= component_count);

    if (filter)
        return component_count - filter->get_filtered_component_count();

    return component_count;
}

t_count t_basic_spectra::get_transaction_count (const t_spectra_filter * filter) const {
    assert(!filter || filter->get_last_transaction() <= transaction_count);

    if (filter)
        return transaction_count - filter->get_filtered_transaction_count();

    return transaction_count;
}

void t_basic_spectra::set_element_count (t_count component_count,
                                         t_count transaction_count) {
    this->component_count = component_count;

    this->transaction_count = transaction_count;

    errors.resize(transaction_count, false);
}

void t_basic_spectra::set_component_count (t_count component_count) {
    set_element_count(component_count, get_transaction_count());
}

void t_basic_spectra::set_transaction_count (t_count transaction_count) {
    set_element_count(get_component_count(), transaction_count);
}

t_error t_basic_spectra::get_error (t_transaction_id transaction) const {
    assert(transaction > 0);
    assert(transaction <= transaction_count);

    return errors[transaction - 1];
}

bool t_basic_spectra::is_error (t_transaction_id transaction) const {
    assert(transaction > 0);
    assert(transaction <= transaction_count);

    return get_error(transaction) > 0.5; // TODO: arbitrary threshold
}

void t_basic_spectra::error (t_transaction_id transaction,
                             t_error set) {
    assert(transaction > 0);
    assert(transaction <= transaction_count);

    errors[transaction - 1] = set;
}
}